locale/inc/leatherman/locale/locale.hpp

```diff
@@ -4,13 +4,85 @@
 */
 #pragma once
 #include <locale>
+#include <vector>
+#include <boost/locale/format.hpp>
 
 namespace leatherman { namespace locale {
 
     /**
      * Gets a locale object for the specified locale id.
      * @param id The locale ID, defaults to a UTF-8 compatible system default.
+     * @param domain The catalog domain to use for i18n via gettext.
+     * @param paths Search paths for localization files.
+     * @return The locale. If a locale for the specified domain already exists, it returns
+     * the same locale until clear_domain is called for that domain.
+     * Throws boost::locale::conv::conversion_error if the system locale is invalid or
+     * the catalog for the specified language can't be used with the system locale encoding.
      */
-    std::locale get_locale(std::string const& id = "");
+    const std::locale& get_locale(std::string const& id = "", std::string const& domain = PROJECT_NAME, std::vector<std::string> const& paths = {PROJECT_DIR});
+
+    /**
+     * Clears the locale for a specific domain.
+     * WARNING: This may invalidate existing references, so only use for testing.
+     * @param domain The catalog domain to clear.
+     */
+    void clear_domain(std::string const& domain = PROJECT_NAME);
+
+    /**
+     * Translate text using the locale initialized by this library.
+     * @param s The string to translate.
+     * @param domain The catalog domain to use for i18n via gettext.
+     * @return The translated string.
+     */
+    std::string translate(std::string const& s, std::string const& domain = PROJECT_NAME);
+
+    /**
+     * A formatter class for translating and formatting text according to a domain.
+     */
+    class formatter
+    {
+        std::string domain_;
+
+        std::string fmt(boost::locale::format& fmt);
+
+    public:
+        /**
+         * Construct a new formatter.
+         * @param domain The domain to use for localization.
+         */
+        formatter(std::string domain = PROJECT_NAME) : domain_(std::move(domain)) {}
+
+        /**
+         * Translate and format.
+         * @param s The format string.
+         * @param args Format arguments.
+         * @return The string generated by translating the format string, then applying the arguments.
+         */
+        template <typename... TArgs>
+        std::string operator()(std::string const& s, TArgs... args)
+        {
+            // Create and apply formatter here, as we want to guarantee the lifetime of the arguments.
+            // boost::locale::format doesn't make copies, and a common gotcha is using temporary arguments
+            // to build up the formatter.
+            boost::locale::format form(boost::locale::translate(s));
+            (void) std::initializer_list<int>{ ((void)(form % args), 0)... };
```

Comment:
It took me a minute to sort through what it was doing. Having to explain it to @whopper helped dramatically in my understanding